### PR TITLE
Debugging in docker

### PR DESCRIPTION
**Pull request by rharris-nitro**
Sets up debugging environments for main production app and testing
- creates debugger (debugpy) in .vscode/launch.json and Taskfile
- divvies up Dockerfile into 2 images main (prod) and dev
- also adds docker-compose implementation

### CODE REVIEW

**Comment by rlasskiy** [comment]
it looks like something wrong with this PR

follow committed is from different PR

[changes <> to []](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/724414dee944e1c6285387d6737c48b8a9d5f4b5)
[created Taskfile and added build, run, and test commands](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/ca59d1bc3d4143a72f8b6d739a9d362ce12ef1c5)
[creates taskfile setup documentation](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/81349c2d94e5cf15e4f8c24e8a3a3e72ef669205)
[moves the task: No tasks with description available. Try --list-all t…](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/f05ecd43f2eb17682c2ca403a89de79afc6e601d)
[includes building and running with task](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/ad9ece0d2e14ce8f07e379676f52c68a77650a3e)
[reformatting headers](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/f461b00f2970db1dd8e0e6d0fbfaf61fd42e6582)
[reformatting headers](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/0ad22d6b2e1aef06245279ff7f858a9cf6bb356d)
[adds global variables to task](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/bd2b74465759390173e48ad51d5d84ea41e44f58)
[gets rid of docker commands section as task has made it easier](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/c49e69d8f793326bbbf660aecacf2ce8482f7e9f)
[gets rid of task setup page](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/f3ab62082b69058285bf8bf25e8c1b966408c9d6)

**Comment by rharris-nitro** [comment]
> it looks like something wrong with this PR
> 
> follow committed is from different PR
> 
> [changes <> to []](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/724414dee944e1c6285387d6737c48b8a9d5f4b5) [created Taskfile and added build, run, and test commands](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/ca59d1bc3d4143a72f8b6d739a9d362ce12ef1c5) [creates taskfile setup documentation](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/81349c2d94e5cf15e4f8c24e8a3a3e72ef669205) [moves the task: No tasks with description available. Try --list-all t…](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/f05ecd43f2eb17682c2ca403a89de79afc6e601d) [includes building and running with task](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/ad9ece0d2e14ce8f07e379676f52c68a77650a3e) [reformatting headers](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/f461b00f2970db1dd8e0e6d0fbfaf61fd42e6582) [reformatting headers](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/0ad22d6b2e1aef06245279ff7f858a9cf6bb356d) [adds global variables to task](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/bd2b74465759390173e48ad51d5d84ea41e44f58) [gets rid of docker commands section as task has made it easier](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/c49e69d8f793326bbbf660aecacf2ce8482f7e9f) [gets rid of task setup page](https://github.com/rharris-nitro/pythonExercise1/pull/9/commits/f3ab62082b69058285bf8bf25e8c1b966408c9d6)

fixed this issue by deleting commits using `git rebase -i`

File: requirements.txt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# To ensure app dependencies are ported from your virtual environment/host machine into your container, run 'pip freeze > requirements.txt' in the terminal to overwrite this file
+
+debugpy==1.8.0
```
Comment (rlasskiy) [inline requirements.txt:3]: image

File: .dockerignore [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+**/docker-compose*
+**/compose*
+**/Dockerfile*
+**/node_modules
+**/npm-debug.log
+**/obj
+**/secrets.dev.yaml
+**/values.dev.yaml
+LICENSE
+README.md
```
Comment (rlasskiy) [inline .dockerignore:28]: ??

File: Dockerfile [hunk 3]
```diff
@@ -4,17 +4,24 @@ FROM python:3.11 AS base
...
 FROM base AS main
 
-    WORKDIR /pythonExercise1
+WORKDIR /app
 
-    COPY app/ app/
+COPY app/ .
+
+ENTRYPOINT ["python"]
+CMD ["app/main.py"]
```
Comment (rlasskiy) [inline Dockerfile:12]: CMD ["main.py"]

File: docker-compose.test.yml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+version: '3.4'
+
+services:
+  my-python-test:
+    build:
+      context: .
+      target: test
+    image: my-python-test
```
Comment (rlasskiy) [inline docker-compose.test.yml:8]: please add possibility to debug the test as well

File: Taskfile.yml [hunk 5]
```diff
@@ -1,22 +1,20 @@
 version: "3"
 
-env:
-  MAIN_CONTAINER: my-python-app
-  TEST_CONTAINER: my-python-test
-
 tasks:
-  build-app:
+  build:
```
Comment (rlasskiy) [inline Taskfile.yml:4]: documentation is reference 
  ```bash
  task build-app
  ```

File: .gitignore [hunk 6]
```diff
@@ -1,2 +1,15 @@
 */__pycache__/
 .DS_Store
+
+.vscode/*
+!.vscode/settings.json
+!.vscode/tasks.json
+!.vscode/launch.json
```
Comment (rlasskiy) [inline .gitignore:7]: `.vscode/launch.json` is missing in this PR

File: Dockerfile [hunk 7]
```diff
@@ -4,17 +4,24 @@ FROM python:3.11 AS base
...
+COPY tests/test_main.py .
 
-    COPY requirements-test.txt .
-    RUN pip install -r requirements-test.txt
+COPY requirements.test.txt .
+RUN pip install -r requirements.test.txt
 
-    CMD ["pytest", "test_main.py"]
+ENTRYPOINT ["pytest"]
+CMD ["test_main.py"]
```
Comment (rlasskiy) [inline Dockerfile:27]: do you need this?
Do you plan to extend this each time and then add a new test file?

File: docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+services:
+  my-python-app:
+    build:
+      context: .
+      target: main-dev
+    image: my-python-app
+    ports:
+      - "5673:5673"
+    command: ${BASE_COMMAND:- main.py}
```
Comment (rlasskiy) [inline docker-compose.yml:11]: during debugging, we debug files which uploaded to the container during `docker build` process
we need to add the possibility to debug files from the working directory

File: .vscode/launch.json [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+      "type": "python",
+      "request": "attach",
+      "host": "localhost",
+      "port": 5673,
+      "pathMappings": [
+        { "localRoot": "${workspaceFolder}/app", "remoteRoot": "/app" }
+      ]
+    }
+  ]
+}
```
Comment (rlasskiy) [inline .vscode/launch.json:14]: you add [.vscode/launch.json](https://github.com/rharris-nitro/pythonExercise1/pull/9/files#diff-bd5430ee7c51dc892a67b3f2829d1f5b6d223f0fd48b82322cfd45baf9f5e945) but your documentation `Setup local environment` not mention what we need install VSCode

File: Taskfile.yml [hunk 10]
```diff
@@ -1,22 +1,20 @@
...
-      - docker run $MAIN_CONTAINER
+      - docker-compose -f docker-compose.yml up
 
   build-test:
     cmds:
-      - docker build -t $TEST_CONTAINER --target test .
+      - docker-compose -f docker-compose.test.yml build
 
-  run-test:
+  up-test:
```
Comment (rlasskiy) [inline Taskfile.yml:18]: you change the name of the task but not update the documentation which is still reference
  ```bash
  task run-test
  ```

File: docs/local-environment.md [hunk 11]
```diff
@@ -97,3 +87,15 @@ test_main.py ..                                                          [100%]
...
+
+### Debugging main<a name="debugging-main"></a>
+
+To debug the main image in VSCode, run the image using task and set USE_DEBUG to true:
+
+```bash
+task up USE_DEBUG=true
+```
+
+The same can be done for the test image.
```
Comment (rlasskiy) [inline docs/local-environment.md:101]: Attach to debugger via VSCode step is missing

File: docs/local-environment.md [hunk 12]
```diff
@@ -50,37 +51,29 @@ brew install go-task/tap/go-task
 
 For more info, visit the [Taskfile website](https://taskfile.dev/)
 
-## Build Docker image and run container using Task<a name="build-docker-image-and-run-container-using-task"></a>
+### Install VSCode<a name="install-vscode"></a>
 
-### Run App<a name="run-app"></a>
+You should open the folder in VSCode in the directory where the repo is cloned.
```
Comment (rlasskiy) [inline docs/local-environment.md:56]: Please add the link to VSCode

File: docs/debugging.md [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+The debugging tool inside the Docker container must listen on a specific port. This port must be exposed and mapped to the host machine in the Docker Compose file or Docker run command. For example, if your application listens on port 5673 for debugging, you need to ensure that this port is exposed and mapped in your Docker configuration.
+
+## Waiting for the Debugger:
+
+The application should be configured to wait for the debugger to attach before it starts executing. This is often achieved by passing a command-line argument to the debugging tool that tells it to wait for a connection. For instance, in a Python application using debugpy, you might use --wait-for-client to make the application wait for the debugger to connect.
+
+## VSCode Configuration:
+
+In VSCode, you need to configure a launch configuration that tells VSCode how to connect to the debugger inside the Docker container. This involves specifying the type of debugger (e.g., Python), the host (usually localhost or 127.0.0.1), and the port number that matches the one configured in the Docker container.
```
Comment (rlasskiy) [inline docs/debugging.md:29]: add link to `.vscode/launch.json`

**Review by rlasskiy — APPROVED** [review]

**Pull request merged**